### ISSUE TITLE
Generate Permanent URIs

### ISSUE THREAD

**Issue by giusepperizzo** [3cixty/knowledgebase#56]
This is related to #48 #49 #50 

Approach: 
- use the deeplink pointing to the resource hosted on the platform  where the instance has been taken
- use https://gist.github.com/giusepperizzo/630d32cc473069497ac1 and assign to ID the deeplink

Output: permanent UUID

Problematic case: deeplink does not exist. Then create a standard procedure to generate a unique identifier for the instance. Beware that the ID must be the same over the different runs (share the protocol).

Workflow: fix this issue, and then apply the fix on the related issues.

Tag @ocorcho @rtroncy for checking.
Tag @jcballesteros @jplu for the due changes.

**Comment by rtroncy** [comment]
I more or less agree, but the description above is still ambiguous, because there is not always a single deeplink for a given source (canonical link vs expanded link; internationalization link; mobile links, etc. just to name a few problems). Therefore, it is unclear on which link the SHA-1 hash should be applied.

For example, for both Last.fm and Eventful, those providers have at least 2 links for each first class objects that I will call *canonical link* and *expanded link*. Hence, for Eventful:
* La Scala: http://eventful.com/milan/venues/V0-001-001814788-2 (canonical URL)
* La Scala: http://eventful.com/milan/venues/la-scala-/V0-001-001814788-2 (expanded URL)
Note that the first link (canonical URL) always 301 Moved Permanently to the later link (expanded URL).

As a second example, Foursquare has numerous links for each venue, e.g.
* La Scala (in French): https://fr.foursquare.com/v/teatro-alla-scala/4b755f6af964a520d0082ee3
* La Scala (in Italian): https://fr.foursquare.com/v/teatro-alla-scala/4b755f6af964a520d0082ee3
* La Scala (on a mobile device): https://m.foursquare.com/v/teatro-alla-scala/4b755f6af964a520d0082ee3 that would redirect to (e.g.) https://fr.foursquare.com/mobile/v/teatro-alla-scala/4b755f6af964a520d0082ee3
* La Scala (canonical link): https://foursquare.com/v/4b755f6af964a520d0082ee3
Note again that the canonical link will always 301 Moved Permanently to the expanded URL (/v/teatro-alla-scala/4b755f6af964a520d0082ee3)

I would argue that we should hash the canonical URL for those sources, but this one is sometimes hard to find.

**Comment by ocorcho** [comment]
We have audited all the UUIDs that we are generating for the sources that @jcballesteros has been transforming. We keep a record of deeplink and UUID assigned, so that whenever we regenerate any new RDF data the UUIDs assigned to any element (place, event) is always the same.

This has an impact in the translations that are being made, since we keep track of those UUIDs as well, so as not to lose translations in the process (which is costly).

Hence we need a solution for the events that @jplu @giusepperizzo @rtroncy have been generating and we have translated (e.g. evensi events), since now the translations are disconnected from teh original UUIDs (this is related to issue #50)

**Comment by giusepperizzo** [comment]
We are in the process to do the same for the data sources under our control (Evensi, Evenful, Last.fm). 
To harmonize the process, we wonder: did you use the same algorithm as it's reported above?

**Comment by rtroncy** [comment]
Assigning @jplu to implement the algorithm described above when generating UUIDs for the rdf-ization of evensi and eventmedia. In both cases, this amounts to SHA-1 hash the canonical link identifying the object in the source. Indicate if you take the http or https link.

**Comment by jplu** [comment]
Yes I take care of that today. But before to start I would have few questions.

1) In lastfm and eventful datasets, for events and places, do you confirm that I take the link provided by the ```owl:sameAs``` property?
2) For the places in lastfm and eventful datasets, the values of some ```owl:sameAs``` property belong to foursquare, dbpedia or here. Do I take them anyway? You can see that with the query:

```
select distinct ?sameAs where {
?s a lode:Event.
{
?s dc:publisher <http://www.eventful.com>.
} UNION {
?s dc:publisher <http://www.last.fm>.
}
?s owl:sameAs ?sameAs
}
```

3) For eventful dataset, some places don't have ```owl:sameAs``` property, so which one do I take instead? You can see that we those two query:

```
select count(distinct ?pl) where {
?s a lode:Event.
?s dc:publisher <http://www.eventful.com> .
?s lode:atPlace ?pl.
?pl owl:sameAs ?sameAs.
}
```

And:

```
select count(distinct ?pl) where {
?s a lode:Event.
?s dc:publisher <http://www.eventful.com> .
?s lode:atPlace ?pl.
}
```

4) How do we do for images (media in general), we take the provided URL?

**Comment by rtroncy** [comment]
1/ NO. You do NOT take the value of the ```owl:sameAs``` property. A particular location can have plenty of sameAs links, e.g. <http://data.linkedevents.org/location/149d7b7a-0d4a-4633-a980-0ff5bbcaf34d> ([description](http://3cixty.eurecom.fr/sparql?default-graph-uri=&query=select+*%0D%0Awhere+{%0D%0A+%3Chttp%3A%2F%2Fdata.linkedevents.org%2Flocation%2F149d7b7a-0d4a-4633-a980-0ff5bbcaf34d%3E+%3Fp+%3Fo%0D%0A}&format=text%2Fhtml&timeout=0&debug=on)) from last.fm has been reconciled with dbpedia, here.net and foursquare. What should do is to take the canonical link for hashing. For last.fm and eventful, it seems to me that you better start from the ```rdfs:seeAlso``` property and alter the string to remove the part after the last / or +.

2/ See above. We are talking about generating the 3cixty UUID of those objects. You need to SHA-1 one link, the canonical link.

3/ See above, ```rdfs:seeAlso```

4/ What do you mean by the provided URL? For Flickr image, you can rely again on the ``rdfs:seeAlso``` link (although there is a ```owl:sameAs``` link too). For Last.fm or Evenful, you just have the  ```ma-ont:locator``` :-( Alternatively, you could use a combination of the thing being illustrated (event id) + the type or the string "image". Why do we have ```ma-ont:Image``` only coming from last.fm, eventful and flickr and not the other sources?

5/ You forgot about UserAccount and Agent (performers) :-)

**Comment by giusepperizzo** [comment]
to complement, hoping to clarify: SHA-1 of the canonical link and then compute from the resulting string the UUID as reported in the https://gist.github.com/giusepperizzo/630d32cc473069497ac1

**Comment by rtroncy** [comment]
yep, this is better by saying !

**Comment by jplu** [comment]
And how do we do when there is much more than one ```rdfs:seeAlso``` property attached to the event ? As show the [query[1]](http://mediamixer.eurecom.fr/sparql?default-graph-uri=&query=select+distinct+%3Fs+%28count%28distinct+%3Fo%29+as+%3Fcount%29+where+%7B%3Fs+a+lode%3AEvent.+%3Fs+rdfs%3AseeAlso+%3Fo+.+%3Fs+dc%3Apublisher+%3Chttp%3A%2F%2Fwww.last.fm%3E%7D+group+by+%3Fs+order+by+DESC%28%3Fcount%29&format=text%2Fhtml&timeout=0&debug=on) for Last.fm and this [one[2]](http://mediamixer.eurecom.fr/sparql?default-graph-uri=&query=select+distinct+%3Fs+%28count%28distinct+%3Fo%29+as+%3Fcount%29+where+%7B%3Fs+a+lode%3AEvent.+%3Fs+rdfs%3AseeAlso+%3Fo+.+%3Fs+dc%3Apublisher+%3Chttp%3A%2F%2Fwww.eventful.com%3E%7D+group+by+%3Fs+order+by+DESC%28%3Fcount%29&format=text%2Fhtml&timeout=0&debug=on) for Eventful.

As show this [query[3]](http://mediamixer.eurecom.fr/sparql?default-graph-uri=&query=select+distinct+%3Fdc+where+%7B%3Fs+a+ma-ont%3AImage.+%3Fs+lode%3Aillustrate+%3Fe.+%3Fe+dc%3Apublisher+%3Fdc%7D&format=text%2Fhtml&timeout=0&debug=on), we have ```ma-ont:Image``` coming from:
*http://www.evensi.com
*http://www.expoincitta.com
*http://www.leonardo-ambrosiana.it
*http://www.eventful.com
*http://www.last.fm

[1]:

```
select distinct ?s (count(distinct ?o) as ?count) where {
?s a lode:Event.
?s rdfs:seeAlso ?o .
?s dc:publisher <http://www.eventful.com>
} group by ?s order by DESC(?count)
```

[2]:

```
select distinct ?s (count(distinct ?o) as ?count) where {
?s a lode:Event.
?s rdfs:seeAlso ?o .
?s dc:publisher <http://www.eventful.com>
} group by ?s order by DESC(?count)
```

[3]:

```
select distinct ?dc where {
?s a ma-ont:Image.
?s lode:illustrate ?e.
?e dc:publisher ?dc
}
```

**Comment by rtroncy** [comment]
That's why you should really not try to patch (with a SPARQL query) and guess which link you should take but instead, come back to the original Eventmedia scraping code when parsing last.fm and eventful, since this ambiguity will be gone in the JSON returned by those APIs.

I picked one random example from last.fm, see this [description](http://3cixty.eurecom.fr/sparql?default-graph-uri=&query=select+*%0D%0Awhere+{%0D%0A+%3Chttp%3A%2F%2Fdata.linkedevents.org%2Fevent%2F05d4d832-3d00-4759-9c73-630fc8984e2f%3E+%3Fp+%3Fo%0D%0A}&format=text%2Fhtml&timeout=0&debug=on), and I'm worried to see that ALL 3cixty properties are duplicated! Why?

**Comment by jplu** [comment]
Yes, only the properties added with the patch are duplicated... I found why, the patches add as well the property to each event coming from the "events" graph. I have to modify the patches to do not alternate the "events" and "places" graphs.

**Comment by jplu** [comment]
I will check the original dump coming from eventmedia, yes. Anyway I planned to make those changes in this script so if the problem doesn't appear during this "extraction" step, it should be ok.

**Comment by giusepperizzo** [comment]
Evensi is now sticking with what proposed above. Commits 4699ab5b9b11479e5f95a7dc3d97e8c69d4b5dc6 and 5a4fd95df10f9f2571b33cf70e5061ed8263e4b2.

@jplu follow up the events with dc:publisher lastfm, eventful
